### PR TITLE
Bwc testclusters all

### PR DESCRIPTION
**Pull request by alpar-t**
This PR switches all the BWC projects to use testclusters. 
Since the new build scripts are more concise it might be easier to review them by lokoing at the new files rather than the diff. 

With this change we are very close to being able to remove ClusterFormationTasks completely.

### CODE REVIEW

File: buildSrc/src/minimumRuntime/java/org/elasticsearch/gradle/Version.java [hunk 1]
```diff
@@ -51,6 +51,10 @@ public static Version fromString(final String s) {
         );
     }
 
+    public static Version current() {
```
Comment (mark-vieira) [inline buildSrc/src/minimumRuntime/java/org/elasticsearch/gradle/Version.java]: Maybe we should be more specific here? The `Version` class is generic, and not only used to model ES versions. We should either rename this to `currentElasticsearch()` or similar or move this to `VersionProperties` itself.

File: qa/mixed-cluster/build.gradle [hunk 2]
```diff
@@ -17,58 +17,88 @@
...
-// It is a single task that can be used to backcompat tests against all versions.
-task bwcTest {
+tasks.register("bwcTest") {
   description = 'Runs backwards compatibility tests.'
   group = 'verification'
 }
 
-for (Version version : bwcVersions.wireCompatible) {
-  String baseName = "v${version}"
+configurations {
```
Comment (mark-vieira) [inline qa/mixed-cluster/build.gradle]: Is there a reason we can't continue to use `RestIntegTestTask` which sets up this configuration and dependency for us?

File: qa/mixed-cluster/build.gradle [hunk 3]
```diff
@@ -17,58 +17,88 @@
...
+      javaHome = project.file(project.ext.runtimeJavaHome)
+    }
   }
 
-  if (project.bwc_tests_enabled) {
-    bwcTest.dependsOn(versionBwcTest)
+  tasks.register("${baseName}#mixedClusterTest", RestTestRunnerTask) {
+    useCluster testClusters."${baseName}"
+    mustRunAfter(precommit)
+    doFirst {
```
Comment (mark-vieira) [inline qa/mixed-cluster/build.gradle]: Do we want to keep all this logging here? I'm not against it, just want to make sure it's not leftover debugging.

File: qa/rolling-upgrade/build.gradle [hunk 4]
```diff
@@ -29,112 +30,102 @@ task bwcTest {
...
-    bwcVersion = version
-    numBwcNodes = 3
-    numNodes = 3
-    clusterName = 'rolling-upgrade'
-    setting 'repositories.url.allowed_urls', 'http://snapshot.test*'
-    if (version.onOrAfter('5.3.0')) {
+      setting 'repositories.url.allowed_urls', 'http://snapshot.test*'
+      setting 'path.repo', "${buildDir}/cluster/shared/repo/${baseName}"
       setting 'http.content_type.required', 'true'
+      javaHome = project.file(project.ext.runtimeJavaHome)
```
Comment (mark-vieira) [inline qa/rolling-upgrade/build.gradle]: Isn't this the default? And if not, should it be?

File: qa/rolling-upgrade/build.gradle [hunk 5]
```diff
@@ -29,112 +30,102 @@ task bwcTest {
...
-    enabled = project.bwc_tests_enabled
-    dependsOn = [upgradedClusterTest]
+    nonInputProperties.systemProperty('tests.rest.cluster', "${-> testClusters."${baseName}".allHttpSocketURI.join(",") }")
+    nonInputProperties.systemProperty('tests.clustername', "${-> testClusters."${baseName}".getName() }")
   }
 
   if (project.bwc_tests_enabled) {
-    bwcTest.dependsOn(versionBwcTest)
+    bwcTest.dependsOn(
+            tasks.register("${baseName}#bwcTest") {
```
Comment (mark-vieira) [inline qa/rolling-upgrade/build.gradle]: nit: can we fix this weird indentation here?

File: x-pack/qa/rolling-upgrade-basic/build.gradle [hunk 6]
```diff
@@ -1,126 +1,113 @@
...
-        // Use all running nodes as seed nodes so there is no race between pinging and the tests
-        { [oldClusterTest.nodes.get(2).transportUri(), oneThirdUpgradedTest.nodes.get(0).transportUri()] })
-
-    Task twoThirdsUpgradedTestRunner = tasks.getByName("${baseName}#twoThirdsUpgradedTestRunner")
-    twoThirdsUpgradedTestRunner.configure {
+    tasks.register("${baseName}#twoThirdsUpgradedTest", RestTestRunnerTask) {
+        dependsOn "${baseName}#oneThirdUpgradedTest"
+        useCluster testClusters."${baseName}"
+        doFirst {
+            testClusters."${baseName}".nextNodeToNextVersion()
```
Comment (mark-vieira) [inline x-pack/qa/rolling-upgrade-basic/build.gradle]: I do tend to agree with Ryan's concerns that this sort of "mutable node" stuff makes it a bit hard to follow. To determine what the state of the cluster is for any given task you have to manually follow the task graph backwards, which is not an easy thing to do.

As we discussed before though, I think this is fine so long as we agree this is a temporary situation until we develop a standard high level model for external cluster testing that supports all the current use cases.

**Review by mark-vieira — APPROVED** [review]
This all generally looks good. As we've talked about, we want to make improvements to the DSL and model here but that can be done later. For me priority #1 is migrating everything to test clusters and turning parallel execution on across the board for all our builds.

File: qa/mixed-cluster/build.gradle (hunk 2 above)
Comment (atorok) [reply qa/mixed-cluster/build.gradle]: Mostly because it also creates a cluster and we want to set that up differently and explicitly here. 
I plan at looking at this at some point. 
Ideally we wouldn't need this and it could be a regular dependency, but we have to make the test runner read stuff off the classpath for that.

File: qa/rolling-upgrade/build.gradle (hunk 4 above)
Comment (atorok) [reply qa/rolling-upgrade/build.gradle]: I wouldn't want to have a plugin rely on some project extension for a default.
I plan to deal with as part of #40531

File: qa/mixed-cluster/build.gradle (hunk 3 above)
Comment (atorok) [reply qa/mixed-cluster/build.gradle]: We need to get the endpoints to make sure the cluster is up before we force an upgrade.  It's not ideal but something we will address with the improvements we are considering.

File: qa/rolling-upgrade/build.gradle (hunk 5 above)
Comment (atorok) [reply qa/rolling-upgrade/build.gradle]: I agree it's weird but it's what IDEA suggests, I prefer to keep it so it doesn't go back and forth as we edit the file.

File: x-pack/qa/rolling-upgrade-basic/build.gradle (hunk 6 above)
Comment (atorok) [reply x-pack/qa/rolling-upgrade-basic/build.gradle]: I agree, we will need a model.

File: buildSrc/src/minimumRuntime/java/org/elasticsearch/gradle/Version.java (hunk 1 above)
Comment (atorok) [reply buildSrc/src/minimumRuntime/java/org/elasticsearch/gradle/Version.java]: I moved the method as suggested

**Comment by atorok** [comment]
@elasticmachine run elasticsearch-ci/2
